### ISSUE TITLE
snap auto refresh lxd 3.15 kill lxcfs service

### ISSUE THREAD

**Issue by fungnaz** [lxc/lxd#6000]
* Distribution: Ubuntu 
 * Distribution version: 16.04.4 LTS
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.4.0-116-generic
   * LXD version: snap 3.15
   * Storage backend in use: zfs

# Issue description
snap auto refresh to LXD 3.15 and found /proc incorrect behavior 

**on container**
root@envoy:~# free -m
Error: /proc must be mounted
  To mount /proc at boot you need an /etc/fstab line like:
      proc   /proc   proc    defaults
  In the meantime, run "mount proc /proc -t proc"

**on host**
when use ps fauxww | grep lxcfs
not found any service running 

**Fix**
on host
1. systemctl status snap.lxd.daemon
2. restart container 

PS: after restart container I found mount proc boot_id, it's new on this release?
none /proc/sys/kernel/random/boot_id tmpfs ro,nosuid,nodev,noexec,relatime,size=492k,mode=755,uid=1000000,gid=1000000 0

**Comment by valcryst** [comment]
Same happend to me on 6 different Hosts and almost 30 Containers.
Is there an option to disable Auto-Refresh so problems like this dont damage
all lxd running hosts ?

I also got dmesg spammed with errors on all affected Hosts.

`check_swap[21117]: segfault at 0 ip 00007f752cfe3aee sp 00007ffea87a6bb0 error 4 in libc-2.23.so[7f752cf76000+1c0000]
`

**Comment by brauner** [comment]
> PS: after restart container I found mount proc boot_id, it's new on this release?
> none /proc/sys/kernel/random/boot_id tmpfs ro,nosuid,nodev,noexec,relatime,size=492k,mode=755,uid=1000000,gid=1000000 0

That's done by liblxc and not lxcfs so it's very unlikely that this would cause any trouble for lxcfs.

**Comment by brauner** [comment]
> Same happend to me on 6 different Hosts and almost 30 Containers.
> Is there an option to disable Auto-Refresh so problems like this dont damage
> all lxd running hosts ?
> 
> I also got dmesg spammed with errors on all affected Hosts.
> 
> `check_swap[21117]: segfault at 0 ip 00007f752cfe3aee sp 00007ffea87a6bb0 error 4 in libc-2.23.so[7f752cf76000+1c0000] `

What's `check_swap`, do you know by any change what tool that is?

**Comment by valcryst** [comment]
> What's `check_swap`, do you know by any change what tool that is?

It's a Nagios Check from the Ubuntu Nagios Plugins Package that also checks /proc and Memory

**Comment by fungnaz** [comment]
> Same happend to me on 6 different Hosts and almost 30 Containers.
> Is there an option to disable Auto-Refresh so problems like this dont damage
> all lxd running hosts ?
> 

snapd has no option to disable the update, 
but I disable by set  proxy feature by put invalid proxy server

snap set system proxy.http="http://127.0.0.1:1111"
snap set system proxy.https="http://127.0.0.1:1111"

or you can disable by Firewall or iptables

**Comment by zsi-vienna** [comment]
Same problem here, adding some observations: 4 hosts were upgraded to 3.15, all unprivileged containers show "Error: /proc must be mounted" when executing "ps -ef". One privileged container has no issue. In one container I also observed this behaviour:
"""
root@scraper1:~# ps -ef
/proc/meminfo: Software caused connection abort
root@scraper1:~# ps -efH
Error: /proc must be mounted
  To mount /proc at boot you need an /etc/fstab line like:
      proc   /proc   proc    defaults
  In the meantime, run "mount proc /proc -t proc"
root@scraper1:~# ps -ef
Error: /proc must be mounted
  To mount /proc at boot you need an /etc/fstab line like:
      proc   /proc   proc    defaults
  In the meantime, run "mount proc /proc -t proc"
"""

**Comment by brauner** [comment]
Can someone provide a coredump of this somehow?

**Comment by fredt34** [comment]
3 hosts impacted, 100s of critical containers - so far the applications inside containers still run, we blocked any update or restart on services.

Additional info:
On rebooted containers (by "sudo reboot" from inside / and / lxc restart container1), "uptime -p" displays the host' uptime.

**Comment by brauner** [comment]
I'm looking into the issue on the LXCFS side.

**Comment by brauner** [comment]
I suspect this is introduced by the cpu virtualization commit.

**Comment by gaetanquentin** [comment]
I had this problem today and i posted on snapcraft forum: 

https://forum.snapcraft.io/t/lxd-snap-refresh-make-containers-hangs-and-disconnect-exec-shells/12419


Really, snap makes things very dangerous for production environment!

**Comment by fredt34** [comment]
If this can help : the problem occurs on 2 hosts out of 3 - those with 8 and 4 cores according to ```cat /proc/cpuinfo | grep processor | wc -l```, while the test host with only 2 cores is not impacted.

**Comment by brauner** [comment]
I did a bisect and I can't seem to reproduce this, i.e. I can't find a bug in lxcfs at the moment.

**Comment by stgraber** [comment]
I'm reverting the snap to LXCFS 3.0.4 for the time being.

**Comment by fredt34** [comment]
Hmmmm that's not good news !

Here is an extract form our log if it can be of any help:

```
cat syslog | grep daemon

Jul 23 09:38:43 int01 systemd[1]: Closed Socket unix for snap application lxd.daemon.
Jul 23 09:38:44 int01 systemd[1]: Stopping Service for snap application lxd.daemon...
Jul 23 09:38:44 int01 lxd.daemon[16444]: => Stop reason is: snap refresh
Jul 23 09:38:44 int01 lxd.daemon[16444]: => Stopping LXD
Jul 23 09:38:44 int01 lxd.daemon[19313]: => LXD exited cleanly
Jul 23 09:38:45 int01 systemd[1]: Stopped Service for snap application lxd.daemon.
Jul 23 09:39:04 int01 systemd[1]: Listening on Socket unix for snap application lxd.daemon.
Jul 23 09:39:05 int01 systemd[1]: Started Service for snap application lxd.daemon.
Jul 23 09:39:05 int01 lxd.daemon[17208]: => Preparing the system
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Loading snap configuration
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Setting up mntns symlink (mnt:[4026532591])
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Setting up kmod wrapper
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Preparing /boot
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Preparing a clean copy of /run
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Preparing a clean copy of /etc
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Setting up ceph configuration
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Setting up LVM configuration
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Rotating logs
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Setting up ZFS (0.7)
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Escaping the systemd cgroups
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Escaping the systemd process resource limits
Jul 23 09:39:05 int01 lxd.daemon[17208]: ==> Disabling shiftfs on this kernel (auto)
Jul 23 09:39:05 int01 lxd.daemon[17208]: => Re-using existing LXCFS
Jul 23 09:39:05 int01 lxd.daemon[17208]: => Starting LXD
Jul 23 09:39:05 int01 lxd.daemon[17208]: t=2019-07-23T09:39:05+0000 lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored."
Jul 23 09:39:06 int01 lxd.daemon[2372]: mount namespace: 8
Jul 23 09:39:06 int01 lxd.daemon[2372]: hierarchies:
Jul 23 09:39:06 int01 lxd.daemon[2372]:   0: fd:   9: hugetlb
Jul 23 09:39:06 int01 lxd.daemon[2372]:   1: fd:  10: rdma
Jul 23 09:39:06 int01 lxd.daemon[2372]:   2: fd:  11: memory
Jul 23 09:39:06 int01 lxd.daemon[2372]:   3: fd:  12: freezer
Jul 23 09:39:06 int01 lxd.daemon[2372]:   4: fd:  13: net_cls,net_prio
Jul 23 09:39:06 int01 lxd.daemon[2372]:   5: fd:  14: devices
Jul 23 09:39:06 int01 lxd.daemon[2372]:   6: fd:  15: blkio
Jul 23 09:39:06 int01 lxd.daemon[2372]:   7: fd:  16: pids
Jul 23 09:39:06 int01 lxd.daemon[2372]:   8: fd:  17: perf_event
Jul 23 09:39:06 int01 lxd.daemon[2372]:   9: fd:  18: cpuset
Jul 23 09:39:06 int01 lxd.daemon[2372]:  10: fd:  19: cpu,cpuacct
Jul 23 09:39:06 int01 lxd.daemon[2372]:  11: fd:  20: name=systemd
Jul 23 09:39:06 int01 lxd.daemon[2372]:  12: fd:  21: unified
Jul 23 09:39:06 int01 lxd.daemon[2372]: lxcfs.c: 105: do_reload: lxcfs: reloaded
Jul 23 09:39:06 int01 lxd.daemon[17208]: => LXD is ready
```
Plus: They are not "real" servers, but openstack machines.

So far the containers run, we didn't touch them. I'd prefer not restarting lxcfs on a host and its containers as long as the problem is not identified...

We've blocked snap updates with the nice "proxy hack".

**Comment by stgraber** [comment]
@fredt34 is that from the system where things are still running properly? I'm not seeing any error in that log, it looks like the normal upgrade logic.

**Comment by stgraber** [comment]
Reproduced the issue, unfortunately it's not as simple as 3.1.1 is broken, it's that 3.0.4 to 3.1.1 while running is broken...

So us reverting would just break more users... the real fix is figure out why the reload causes a crash.

**Comment by fredt34** [comment]
> @fredt34 is that from the system where things are still running properly? I'm not seeing any error in that log, it looks like the normal upgrade logic.

No, this is from the "big guy" with 8 cores. One container was rebooted inside and gets host's uptime.

**Comment by stgraber** [comment]
For those affected, there's sadly no magic way out of this, your best bet is:
 - Reload LXD: sudo systemctl reload snap.lxd.daemon
 - Check that lxcfs is running: ps aux | grep lxcfs
 - Restart any container that you don't mind restarting now
 - For any container that you can't restart now, go in and unmount lxcfs
    - lxc exec NAME bash
    - grep lxcfs /proc/mounts
    - umount PATH for all of the above

**Comment by stgraber** [comment]
@fredt34 can you confirm that your unaffected machine is also running revision 11320 (per `snap list`)?

**Comment by stgraber** [comment]
I myself have a number of affected systems, though I don't think it's a simple matter of CPU count as an unaffected one is running 8 cores while an affected one is running 48 cores while another affected one only has 4 cores.

**Comment by fredt34** [comment]
Yes, our unaffected tst02 runs 11320 - both 2 other systems also.

(I hoped that this 2-4-8 would be a possible clue towards the "cpu virtualisation" track)

**Comment by stgraber** [comment]
Yeah, it sadly isn't. We've also been failing at causing it on straight upgrade from 3.0.4 to 3.1.1...
My best guess is that it depends to the data in flight at the time of the reload, which makes this rather terrible to track down, but also explains why none of our CI caught it...

**Comment by stgraber** [comment]
I'm currently doing some more checks to ensure that when lxcfs is asked to reload that it reloads the right binary, then if that looks all correct, I'll try to cause some varying 3.0.4 to 3.1.1 updates to see if I can trigger the issue somehow.

**Comment by stgraber** [comment]
Moving this issue over to lxcfs
